src/event-log.c

```diff
@@ -0,0 +1,398 @@
+
+	/* Close open stream */
+	g_clear_pointer(&logger->logstream, g_object_unref);
+
+	/* iterate through list of found existing files and delete the first
+	 * one(s) and move the others */
+	for (guint file_num = logger->maxfiles; file_num > 0; file_num--) {
+		g_autofree gchar *from_file = g_strdup_printf("%s.%d", logger->filename, file_num -1);
+		g_autofree gchar *to_file = g_strdup_printf("%s.%d", logger->filename, file_num);
+
+		/* skip non-existing files */
+		if (!g_file_test(from_file, G_FILE_TEST_EXISTS))
+			continue;
+
+		if (g_rename(from_file, to_file) == -1) {
+			int err = errno;
+			g_warning("Failed to rotate log file: %s", g_strerror(err));
+		}
+		continue;
+	}
+
+	/* rotate current log file as next .1 file */
+	rotfile = g_strdup_printf("%s.1", logger->filename);
+	if (g_rename(logger->filename, rotfile) == -1) {
+		int err = errno;
+		g_warning("Failed to rotate log file: %s", g_strerror(err));
+		return;
+	}
+	logger->filesize = 0;
+
+	/* re-open log file */
+	logfile = g_file_new_for_path(logger->filename);
+	logger->logstream = g_file_append_to(logfile, G_FILE_CREATE_NONE, NULL, &ierror);
+	if (!logger->logstream) {
+		g_warning("Failed to open log file for appending: %s", ierror->message);
+		return;
+	}
+}
+
+static void event_log_writer_file(REventLogger* logger, const GLogField *fields, gsize n_fields)
+{
+	GError *ierror = NULL;
+	g_autofree gchar *formatted = NULL;
+	g_autofree gchar *output = NULL;
+	gsize written = 0;
+
+	g_assert_nonnull(logger->filename);
+
+	if (!logger->logstream) {
+		g_autoptr(GFile) logfile = NULL;
+		logfile = g_file_new_for_path(logger->filename);
+		logger->logstream = g_file_append_to(logfile, G_FILE_CREATE_NONE, NULL, &ierror);
+		if (!logger->logstream) {
+			g_warning("Failed to open log file for appending: %s", ierror->message);
+			return;
+		}
+	}
+
+	switch (logger->format) {
+		case R_EVENT_LOGFMT_READABLE:
+			formatted = event_log_format_fields_readable(G_LOG_LEVEL_MESSAGE, fields, n_fields, TRUE);
+			break;
+		case R_EVENT_LOGFMT_READABLE_SHORT:
+			formatted = event_log_format_fields_readable(G_LOG_LEVEL_MESSAGE, fields, n_fields, FALSE);
+			break;
+		case R_EVENT_LOGFMT_JSON:
+#if ENABLE_JSON
+			formatted = event_log_format_fields_json(G_LOG_LEVEL_MESSAGE, fields, n_fields, FALSE);
+#else
+			g_error("Compiled without JSON support");
+#endif
+			break;
+		case R_EVENT_LOGFMT_JSON_PRETTY:
+#if ENABLE_JSON
+			formatted = event_log_format_fields_json(G_LOG_LEVEL_MESSAGE, fields, n_fields, TRUE);
+#else
+			g_error("Compiled without JSON support");
+#endif
+			break;
+		default:
+			g_error("Unknown log format");
+	}
+	output = g_strdup_printf("%s\n", formatted);
+
+	/* Once we know how much to write, we can use this information for trimming */
+	rotate_logfiles(logger, strlen(output));
+
+	if (!g_output_stream_write_all(G_OUTPUT_STREAM(logger->logstream), output, strlen(output), &written, NULL, &ierror)) {
+		g_warning("Failed to write log file: %s", ierror->message);
```

Comment:
Hmm, this will keep generating `g_warnings` for any further message. Perhaps we should have a `broken` flag per logger and drop any messages for broken loggers?

Reply:
Added `broken` flag now (+test).